torch/_dynamo/symbolic_convert.py

```diff
@@ -774,27 +775,106 @@ def maybe_has_backedge(self):
-        if not hasattr(self, "_cell_and_freevars"):
-            self._cell_and_freevars = tuple(
-                self.code_options["co_cellvars"] or []
-            ) + tuple(self.code_options["co_freevars"] or [])
+    def cellvars(self):
+        if not hasattr(self, "_cellvars"):
+            self._cellvars = tuple(self.code_options["co_cellvars"] or [])
+            # An inlined function might depend on the cellvar of the parent
+            # function. So, recursively obtain parent cellvars.
+            if isinstance(self, InliningInstructionTranslator):
+                self._cellvars += self.parent.cellvars()
+        return self._cellvars
 
+    def freevars(self):
+        if not hasattr(self, "_freevars"):
+            self._freevars = tuple(self.code_options["co_freevars"] or [])
             # An inlined function might depend on the freevar of the parent
-            # function. So, recursively obtain parent cell and freevars.
+            # function. So, recursively obtain parent freevars.
             if isinstance(self, InliningInstructionTranslator):
-                self._cell_and_freevars += self.parent.cell_and_freevars()
+                self._freevars += self.parent.freevars()
+        return self._freevars
+
+    def cell_and_freevars(self):
+        if not hasattr(self, "_cell_and_freevars"):
+            self._cell_and_freevars = self.cellvars() + self.freevars()
         return self._cell_and_freevars
 
     def prune_dead_locals(self):
         reads = livevars_analysis(self.instructions, self.current_instruction)
         # implicit use by super()
         # reads = reads | {"__class__"}
         # output variables?
-        reads = reads | set(self.cell_and_freevars())
+        reads = reads | set(self.freevars())
+
+        # First prune the non-cell local vars, this allows us to prune more cell
+        # local vars later on (e.g., if we manage to prune a
+        # `NestedUserFunctionVariable` that makes use of some cell locals).
+        cellvars = set(self.cellvars())
         self.symbolic_locals = {
-            k: v for k, v in self.symbolic_locals.items() if k in reads
+            k: v for k, v in self.symbolic_locals.items() if k in cellvars or k in reads
         }
+
+        # Then prune the cell locals.
+        #
+        # Note that we keep certain cell locals, because the current mechanism
+        # for codegen closure initialization for nested function creation is:
+        # 1. `NestedUserFunctionVariable` codegen assumes its closure has been
+        #    initialized properly by its creater, i.e., the tuple of cells will
+        #    be populated with correct content before the function is used.
+        # 2. In the fast path of `OutputGraph::compile_subgraph`, we populate
+        #    the tuple of cells _after_ emitting the `MAKE FUNCTION` bytecode,
+        #    via `STORE DEREF`; these `STORE DEREF` are generated partly based
+        #    on the current `symbolic_locals`.
```

Comment:
I thought about another approach where we emit `STORE_DEREF` as part of `NestedUserFunctionVariable::reconstruct`, and get rid of the special handling of `cellvars` here; I might look into that more if I can't fix the existing test failures.

Reply:
I'm starting to think there might be a better solution with this approach, by replacing uses of `ClosureVariable` with `NewCellVariable`. It feels related to #136814 as well.